### PR TITLE
added report system

### PR DESCRIPTION
**Pull request by Olde7325**
This pull requests adds a report system, enhancing the native report command with additional functionalities.

## Summary
These CCs allow users to create a report system with the ability for their server members to request cancellation/nullification of their reports and adding some functionalities for staff utilizing reactions. All necessary information is arranged in an embed which is edited accordingly.

## Documentation
The system should be documented well enough for users with basic CC knowledge on how to add and set up this package.
Please take the time to read the documentation found in the README.md, as this is the crucial part for user accessibility and suggest possible changes.

## Stability
The code was tested thoroughly and should be stable, as long as the end user only changes the parts which are intended for changing. It might be slow from time to time, this is given by the amount of information YAGPDB puts together in the report message. 

Everything else is explained in the README file.

## Status
- [x] Code changes have been tested on an instance of YAGPDB, or there are no code changes
- [x] I have read and followed the [contribution guide](../CONTRIBUTING.md)

---
This is my first PR to this repository. Thanks in advance!

### CODE REVIEW

File: moderation/reports/cancelReport.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    Created by: Olde7325 <https://github.com/Olde7325>
+*/}}
+
+{{/*ACTUAL CODE*/}}
+{{$p := index (reFindAllSubmatches `.*?: \x60(.*)\x60\z` (execAdmin "prefix")) 0 1}}
+{{$Escaped_Prefix := reReplace `[\.\[\]\-\?\!\\\*\{\}\(\)\|]` $p `\${0}`}}
+{{if not (reFind (print `\A` $Escaped_Prefix `|<@!?204255221017214977>`) .Message.Content)}}
+{{$response:= sendMessageRetID nil "Did not set regex to match Server Prefix!"}}{{deleteTrigger}}
+{{deleteMessage nil $response}}{{else}}
+{{if not (ge (len .CmdArgs) 3)}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:19]:
```suggestion
{{if lt (len .CmdArgs) 3}}
```

File: moderation/reports/cancelReport.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        {{end}}
+    {{end}}
+    {{$user := index (reFindAllSubmatches `\A<@!?(\d{17,19})>` $r.Description) 0 1|toInt|userArg}}
+    {{$userKey := (dbGet .User.ID "key").Value|str}}
+    {{if eq $user.ID .User.ID}}
+            {{if eq "used" $userKey}}
+                {{$response := sendMessageRetID nil "Your latest report was already cancelled!"}}
+                {{deleteMessage nil $response}}
+            {{else}}
+            {{if eq (index .CmdArgs 1|str) $userKey}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:39]:
```suggestion
            {{if eq (index .CmdArgs 1) $userKey}}
```

File: moderation/reports/cancelReport.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                {{if ge (len .CmdArgs) 3}}
+                    {{$reason := joinStr " " (slice .CmdArgs 2)}}
+                    {{with $r}}
+                        {{.Set "Author" (sdict "name" (printf "%s (ID %d)" $user.String $user.ID) "icon_url" ($user.AvatarURL "256"))}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:43]:
```suggestion
                        {{.Set "Author" (sdict "name" (printf "%s (ID %d)" $user $user.ID) "icon_url" ($user.AvatarURL "256"))}}
```

File: moderation/reports/cancelReport.go.tmpl [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        {{if eq (kindOf $v true) "struct"}}
+            {{$r.Set $k (structToSdict $v)}}
+        {{end}}
+    {{end}}
+    {{$user := index (reFindAllSubmatches `\A<@!?(\d{17,19})>` $r.Description) 0 1|toInt|userArg}}
+    {{$userKey := (dbGet .User.ID "key").Value|str}}
+    {{if eq $user.ID .User.ID}}
+            {{if eq "used" $userKey}}
+                {{$response := sendMessageRetID nil "Your latest report was already cancelled!"}}
+                {{deleteMessage nil $response}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:37]:
```suggestion
                Your latest report was already cancelled! {{ deleteResponse }}
```

File: moderation/reports/cancelReport.go.tmpl [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    Created by: Olde7325 <https://github.com/Olde7325>
+*/}}
+
+{{/*ACTUAL CODE*/}}
+{{$p := index (reFindAllSubmatches `.*?: \x60(.*)\x60\z` (execAdmin "prefix")) 0 1}}
+{{$Escaped_Prefix := reReplace `[\.\[\]\-\?\!\\\*\{\}\(\)\|]` $p `\${0}`}}
+{{if not (reFind (print `\A` $Escaped_Prefix `|<@!?204255221017214977>`) .Message.Content)}}
+{{$response:= sendMessageRetID nil "Did not set regex to match Server Prefix!"}}{{deleteTrigger}}
+{{deleteMessage nil $response}}{{else}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:18]:
```suggestion
Did not set regex to match Server Prefix! {{deleteTrigger}}{{deleteResponse}}
{{else}}
```

File: moderation/reports/cancelReport.go.tmpl [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                    {{end}}
+                    {{editMessage $rL $rID (complexMessageEdit "embed" $r)}}
+                    Cancellation requested, have a nice day!
+                    {{deleteAllMessageReactions $rL $rID}}
+                    {{addMessageReactions $rL $rID "🚫" "✅" "⚠️"}}
+                    {{dbSet .User.ID "key" "used"}}
+                {{end}}
+            {{else}}
+                {{$response := sendMessageRetID nil "Invalid key provided!"}}
+                {{deleteMessage nil $response}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:57]:
```suggestion
                Invalid key provided!
                {{deleteResponse}}
```

File: moderation/reports/cancelReport.go.tmpl [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            {{end}}
+        {{end}}
+        {{else}}
+            {{$response := sendMessageRetID nil "You are not the author of this report!"}}
+            {{deleteMessage nil $response}}
```
Comment (Jo3-L) [inline moderation/reports/cancelReport.go.tmpl:62]:
```suggestion
            You are not the author of this report!
            {{deleteResponse}}
```

File: moderation/reports/customReport.go.tmpl [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+{{/*ACTUAL CODE*/}}
+{{$p := index (reFindAllSubmatches `.*?: \x60(.*)\x60\z` (execAdmin "prefix")) 0 1}}
+{{$Escaped_Prefix := reReplace `[\.\[\]\-\?\!\\\*\{\}\(\)\|]` $p `\${0}`}}
+{{if not (reFind (print `\A` $Escaped_Prefix `|<@!?204255221017214977>`) .Message.Content)}}
+{{$response:= sendMessageRetID nil "Did not set regex to match Server Prefix!"}}{{deleteTrigger}}
+{{deleteMessage nil $response}}{{else}}
+{{if and .CmdArgs (lt (len .CmdArgs) 2)}}
+    {{if eq (index .CmdArgs 0) "dbSetup"}}
+        {{if (in (split (index (split (exec "viewperms") "\n") 2) ", ") "ManageServer")}}
+            {{dbSet 2000 "reportSettings" (sdict "reportLog" (str $REPORT_LOG
```
Comment (Jo3-L) [inline moderation/reports/customReport.go.tmpl:30]: Saving `sdict` to database preserves the original type. So you can consider not stringifying the IDs and using them directly.. just a thought.

File: moderation/reports/customReport.go.tmpl [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        ) "reportDiscussion" (str $REPORT_DISCUSSION))}}
+            {{dbSet 2000 "ReportNo" 0}}
+            {{sendMessage nil "**Database primed, report count reset, system is ready to use!**"}}
```
Comment (Jo3-L) [inline moderation/reports/customReport.go.tmpl:33]:
```suggestion
            **Database primed, report count reset, system is ready to use!**
```

File: moderation/reports/customReport.go.tmpl [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        {{else}}
+            {{sendMessage nil "You do not have permission to use this command!"}}
```
Comment (Jo3-L) [inline moderation/reports/customReport.go.tmpl:35]:
```suggestion
            You do not have permission to use this command!
```

File: moderation/reports/customReport.go.tmpl [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    {{sendMessage nil (printf "```%s <User:Mention/ID> <Reason:Text>``` \n Not enough arguments passed." .Cmd)}}
+    {{end}}
+{{else if not .CmdArgs}}
+    {{sendMessage nil (printf "```%s <User:Mention/ID> <Reason:Text>``` \n Not enough arguments passed." .Cmd)}}
+{{else}}
+    {{$user := userArg (index .CmdArgs 0)}}
+    {{if eq $user.ID .User.ID}}
+        {{$silly := sendMessageRetID nil "You can't report yourself, silly."}}
+        {{deleteTrigger}}
+        {{deleteMessage nil $silly}}
```
Comment (Jo3-L) [inline moderation/reports/customReport.go.tmpl:47]:
```suggestion
        You can't report yourself, silly. {{deleteTrigger}} {{deleteResponse}}
```

File: moderation/reports/customReport.go.tmpl [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    {{else}}
+        {{$secret := adjective}}
+        {{$logs250 := execAdmin "log" "250"}}
```
Comment (Jo3-L) [inline moderation/reports/customReport.go.tmpl:50]:
```suggestion
        {{$logs250 := execAdmin "log 250"}}
```

**Review by Jo3-L — CHANGES_REQUESTED** [review]
initial review, just general code quality things, overall looks fine.

**Comment by Olde7325** [comment]
I will do these changes in one commit instead of committing each change individually, if that is fine?

**Review by jo3-l2 — APPROVED** [review]

**Comment by jo3-l2** [comment]
lgtm, thanks!

**Comment by Olde7325** [comment]
Thank you very much, as well! Have a happy new year 😄

**Pull request merged**